llvm/test/SYCLLowerIR/esimd_lower_intrins.ll

```diff
@@ -170,6 +170,17 @@ define dso_local spir_kernel void  @FUNC_30() !sycl_explicit_simd !1 {
 ; CHECK-NEXT: ret void
 }
 
+define dso_local spir_kernel void  @FUNC_31() !sycl_explicit_simd !1 {
+; CHECK: define dso_local spir_kernel void  @FUNC_31() !sycl_explicit_simd !1
+  %call = call spir_func i64 @_Z27__spirv_LocalInvocationId_xv()
+; CHECK-NEXT: %call.esimd = call <3 x i32> @llvm.genx.local.id.v3i32()
+; CHECK-NEXT: %local_id.x = extractelement <3 x i32> %call.esimd, i32 0
+; CHECK-NEXT: %local_id.x.cast.ty = zext i32 %local_id.x to i64
```

Comment:
Please, consider following existing design for lowering functions like SPIR-V built-ins to device specific intrinsics by linking corresponding device library with SPIR-V functions implementation (see NVPTX implementation for details).
This will allow you to remove most of `LowerESIMD` pass (including the buggy part).

Reply:
> The problem with NVPTX approach as well as with using existing clang built-in infra is that it is hard to support the "C++" intrinsics mechanism with those. 

I'm not sure why this mechanism is needed. `translateSpirvIntrinsic` doesn't require it.

BTW, why do we need this in the first place? According to my understanding these SPIR-V built-ins are lowered to valid SPIR-V, which should be supported by the back-end.

> Once C++ intrinsics is supported in clang built-in infra - yes, this pass will simplify.
> 
> What buggy part do you refer to?

The one fixed by this patch.

Reply:
I see. You refer to SPIRV translation part. Agree - this could be done similar to NVPTX. 

> According to my understanding these SPIR-V built-ins are lowered to valid SPIR-V, which should be supported by the back-end.

Not quite. The vector BE does not support SPIRV intrinsics, as it is not a SIMT BE. But with subgroup size = 1 restriction it is doable really. I thought about this as a future step.

Reply:
> > According to my understanding these SPIR-V built-ins are lowered to valid SPIR-V, which should be supported by the back-end.
> 
> Not quite. The vector BE does not support SPIRV intrinsics, as it is not a SIMT BE. But with subgroup size = 1 restriction it is doable really. I thought about this as a future step.

To be precise, this functions are represented in SPIR-V as a variable with decoration and I think it must be supported by any back-end. AFAIK, there are a lot of ESIMD specific patches landed to SPIR-V translator and it seems to the right place to lower this functionality.
SPIR-V consumer by design is supposed to lower standard SPIR-V instructions to HW specific intructions/intrinsics. 
If I understand it correctly ESIMD extension implementation includes these two parts:
- LLVM pass replaces SPIR-V built-in functions to GEN specific intrinsics in FE/ME.
- SPIR-V translator is enhanced to preserve undefined GEN specific intrinsics in SPIR-V format.

According to my understanding these are not required if we move lowering of SPIR-V standard instructions to SPIR-V consumer.

NOTE: I'm talking about standard SPIR-V functionality only.

Reply:
> SPIR-V consumer by design is supposed to lower standard SPIR-V instructions to HW specific intructions/intrinsics.

That is true. But ESIMD back-end by design wasn't a "SPIRV consumer" and glue passes are still required to feed it LLVM IR resulting from SPIRV translation. Also ESIMD BE can't consume arbitrary SPIRV with Kernel capability - there are a number of restrictions. E.g. ESIMD BE does not have a concept of workitem, a central one in SIMT world. My point is that there are still design options which need to be discussed with the ESIMD BE devs.

Reply:
> But ESIMD back-end by design wasn't a "SPIRV consumer" and glue passes are still required to feed it LLVM IR resulting from SPIRV translation. Also ESIMD BE can't consume arbitrary SPIRV with Kernel capability - there are a number of restrictions. E.g. ESIMD BE does not have a concept of workitem, a central one in SIMT world. My point is that there are still design options which need to be discussed with the ESIMD BE devs.

If SPIR-V consumption is not designed, what is the point in using SPIR-V then? 
It looks like using LLVM IR as exchange format in ESIMD mode would be much easier and doesn't require so much hacks.

Reply:
> If SPIR-V consumption is not designed, what is the point in using SPIR-V then?
IR stability. 

> doesn't require so much hacks.
There can be parts which can be improved, but I don't see how this justifies calling it hacks.